avro-converter/src/test/java/io/confluent/connect/avro/AvroDataTest.java

```diff
@@ -90,6 +93,109 @@
 
   // Connect -> Avro
 
+  @Test
+  public void testCyclicalAvroSchema() {
+    AvroDataConfig avroDataConfig = new AvroDataConfig.Builder()
+        .with(AvroDataConfig.CONNECT_META_DATA_CONFIG, false)
+        .with(AvroDataConfig.ENHANCED_AVRO_SCHEMA_SUPPORT_CONFIG, false)
+        .build();
+    AvroData listAvroData = new AvroData(avroDataConfig);
+    String linkedListAvroSchema =
+        "{\"type\": \"record\",\"name\": \"linked_list\",\"fields\" : "
+            + "[{\"name\": \"value\", \"type\": \"long\"},"
+            + "{\"name\": \"next\", \"type\": [\"null\", \"linked_list\"],\"default\" : null}]}";
+    org.apache.avro.Schema.Parser avroParser = new org.apache.avro.Schema.Parser();
+    org.apache.avro.Schema avroSchema = avroParser.parse(linkedListAvroSchema);
+
+    GenericRecord next = new GenericRecordBuilder(avroSchema)
+        .set("value", 2l)
+        .set("next", null)
+        .build();
+
+    GenericRecord headNode = new GenericRecordBuilder(avroSchema)
+        .set("value", 3l)
+        .set("next", next)
+        .build();
+
+    SchemaAndValue schema = listAvroData.toConnectData(avroSchema, headNode);
+    GenericRecord genericRecord = (GenericRecord) listAvroData.fromConnectData(
+        schema.schema(), schema.value());
+
+    assertEquals(headNode, genericRecord);
+
+  }
+
+  @Test
+  public void testGraph() {
+
+    AvroDataConfig avroDataConfig = new AvroDataConfig.Builder()
+        .with(AvroDataConfig.CONNECT_META_DATA_CONFIG, false)
+        .with(AvroDataConfig.ENHANCED_AVRO_SCHEMA_SUPPORT_CONFIG, false)
+        .build();
+    AvroData graphAvroData = new AvroData(avroDataConfig);
+    org.apache.avro.Schema.Parser avroParser = new org.apache.avro.Schema.Parser();
+
+    String graphAvroSchema = "{\"type\": \"record\",\"name\": \"Users\",\"fields\" : [{\"name\": " +
+        "\"name\", \"type\": \"string\"},{\"name\": \"friends\", \"type\" : [ \"null\", " +
+        "{\"type\": \"array\", \"items\":\"Users\"}], \"default\" : null}]}";
+
+    org.apache.avro.Schema graphSchema = avroParser.parse(graphAvroSchema);
+    GenericRecord friend1 = new GenericRecordBuilder(graphSchema)
+        .set("name", "Person A")
+        .build();
+    GenericRecord friend2 = new GenericRecordBuilder(graphSchema)
+        .set("name", "Person A")
+        .build();
+
+    org.apache.avro.Schema arraySchema = graphSchema.getField("friends").schema().getTypes().get(1);
+    GenericData.Array friends = new GenericData.Array(arraySchema, Arrays.asList(friend1, friend2));
+    GenericRecord person = new GenericRecordBuilder(graphSchema)
+        .set("name", "Person C")
+        .set("friends", friends)
```

Comment:
I would try nesting 2 levels deep just to further validate the recursive behavior. for example, move the extraction of `arraySchema` up next to the parsing of the original schema (and maybe rename to clearer `friendsListSchema` or something like that) and make `friend2` have `friend1` as a friend as well. then validate we see the full nested version serialized.